### ISSUE TITLE
ASAN heap use after free in ClientTest.TestDeleteCDCStreamAsync

### ISSUE THREAD

**Issue by mbautin** [yugabyte/yugabyte-db#10566]
Observing when running the test in a loop:

```
./yb_build.sh asan --clang12 --cxx-test client_client-test --gtest_filter ClientTest.TestDeleteCDCStreamAsync -n 1000 --tp 1
```

https://gist.githubusercontent.com/mbautin/3b0e2fef44d01ad20236e98efc9a3c91/raw

Stack trace of aborted Raft operation:

https://gist.github.com/mbautin/87c6a33cfb7e297efdd8dd83246fa5dd/raw

**Comment by mbautin** [comment]
```
void InboundCall::Serialize(boost::container::small_vector_base<RefCntBuffer>* output) {
  size_t old_size = output->size();
  DoSerialize(output);
  if (rpc_method_metrics_ && rpc_method_metrics_->response_bytes) {  // <-- ASAN issue happens here.
    auto response_size = 0;
    for (size_t i = old_size; i != output->size(); ++i) {
      response_size += (*output)[i].size();
    }
    if (response_size) {
      rpc_method_metrics_->response_bytes->IncrementBy(response_size);
    }
  }
}
```

It looks like rpc_method_metrics_ points to deallocated memory after MasterServiceImpl has been deleted.

**Comment by mbautin** [comment]
Similar stack traces observed with in a different run:

https://gist.githubusercontent.com/mbautin/7ab3331ceed0c2ad3bfbb0f32f430e38/raw

https://gist.githubusercontent.com/mbautin/a268464fdfc7e1469bcc22e8075c5a7d/raw

**Comment by bmatican** [comment]
as per @spolitov , this recent commit (point 5) should have fixed this: f379f265d24a0144b6b56c7eedeae07ed42814d7
